src/Groups/tom.jl

```diff
@@ -0,0 +1,613 @@
+#############################################################################
+##
+##  tables of marks in Oscar
+##
+abstract type GroupTableOfMarks end
+
+"""
+    GAPGroupTableOfMarks <: GroupTableOfMarks
+
+This is the type of tables of marks that can delegate
+tasks to an underlying table of marks object in the GAP system
+(field `GAPTable`).
+
+A group can (but need not) be stored in the field `group`.
+If it is available then also the field `isomorphism` is available,
+its value is a bijective map from the `group` value to a group in GAP.
+
+Objects of type `GAPGroupTableOfMarks` support [`get_attribute`](@ref).
+"""
+@attributes mutable struct GAPGroupTableOfMarks <: GroupTableOfMarks
+  GAPTable::GapObj  # the GAP table of marks object
+  group::Union{GAPGroup, FinGenAbGroup}    # the underlying group, if any
+  isomorphism::Map  # isomorphism from `group` to a group in GAP
+
+  function GAPGroupTableOfMarks(G::Union{GAPGroup, FinGenAbGroup}, tom::GapObj, iso::Map)
+    return new(tom, G, iso)
+  end
+
+  function GAPGroupTableOfMarks(tom::GapObj)
+    # group and isomorphism are left undefined
+    return new(tom)
+  end
+end
+
+# access to field values via functions
+GapObj(tom::GAPGroupTableOfMarks) = tom.GAPTable
+
+function group(tom::GAPGroupTableOfMarks)
+  @req isdefined(tom, :group) "table of marks stores no group"
+  return tom.group
+end
+
+# If the table of marks stores an Oscar group `G` then
+# an isomorphism from `G` to a GAP group is stored in the field `isomorphism`.
+# Different kinds of isomorphisms can be stored,
+# depending on the type of the stored group,
+# like for character tables.
+#
+function isomorphism_to_GAP_group(tom::GAPGroupTableOfMarks)
+  @req isdefined(tom, :group) "table of marks stores no group"
+  @assert isdefined(tom, :isomorphism) "table of marks with group must store also an isomorphism"
+  return tom.isomorphism
+end
+
+
+"""
+    table_of_marks(G::GAPGroup)
+
+Return the table of marks of the finite group `G`.
+
+# Examples
+```jldoctest
+julia> show(stdout, MIME("text/plain"), table_of_marks(symmetric_group(3)))
+Table of marks of Sym(3)
+
+1: 6      
+2: 3 1    
+3: 2 . 2  
+4: 1 1 1 1
+```
+"""
+function table_of_marks(G::Union{GAPGroup, FinGenAbGroup})
+  return get_attribute!(G, :table_of_marks) do
+    iso = isomorphism_to_GAP_group(G)
+    gaptom = GAPWrap.TableOfMarks(codomain(iso))
+    return GAPGroupTableOfMarks(G, gaptom, iso)
+  end
+end
```

Comment:
To make this type stable and also a bit more concise, how about using the `@attr` property:
```suggestion
@attr GAPGroupTableOfMarks function table_of_marks(G::Union{GAPGroup, FinGenAbGroup})
  iso = isomorphism_to_GAP_group(G)
  gaptom = GAPWrap.TableOfMarks(codomain(iso))
  return GAPGroupTableOfMarks(G, gaptom, iso)
end
```

Reply:
O.k. (as a continuation of #3927)